include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -778,6 +778,33 @@ class FEEvaluationBase
    * MatrixFree object was given at initialization.
    */
   mutable std::vector<types::global_dof_index> local_dof_indices;
+
+  /**
+   * A temporary data structure for the Jacobian information necessary if
+   * the reinit() function is used that allows to construct user batches.
+   */
+  AlignedVector<Tensor<2, dim, VectorizedArrayType>> jacobian_data;
+
+  /**
+   * A temporary data structure for the Jacobian determinant necessary if
+   * the reinit() function is used that allows to construct user batches.
+   */
+  AlignedVector<VectorizedArrayType> J_value_data;
+
+  /**
+   * A temporary data structure for the gradients of the inverse Jacobian
+   * transformation necessary if the reinit() function is used that allows
+   * to construct user batches.
+   */
+  AlignedVector<
+    Tensor<1, dim *(dim + 1) / 2, Tensor<1, dim, VectorizedArrayType>>>
```

Comment:
All of these MF patches are way out of my league, not knowing this code at all. But can you explain where the `d*(d+1)/2` comes from? Is it related to `SymmetricTensor<2,dim>::n_independent_elements`?

Reply:
> All of these MF patches are way out of my league, not knowing this code at all. But can you explain where the d*(d+1)/2 comes from? Is it related to SymmetricTensor<2,dim>::n_independent_elements?

Yes. This is the same. `dim *(dim + 1) / 2` is used since recently in MF and always spelled out. However, at another occasion I already suggested that we should give a more useful name these instances. Why not use `SymmetricTensor::n_independent_components` in these instances? @kronbichler @bergbauer What do you think? I could do this a follow-up PR!

Reply:
Indeed, this is the number of independent components in a symmetric tensor. (The problem I saw when this was introduced in 2013/2014 or even the original code in 2011 that I did not like the overly long lines that we got as a result when not using any local typedef. Not that I always agree with clang-format, but at least it forces code back into the 80 column limit.)

Reply:
I also think it will get a bit lengthy without a `typedef` but it is definitely more descriptive. We should improve that!

Reply:
Thinking about it, why do we not simply use `SymmetricTensor`?

Reply:
We could also introduce a rank 3 Tensor with defined symmetry of two ranks. Just some thoughts.

Reply:
When we made the original design in 2010/2011, `SymmetricTensor` was so slow that each access induced around 20-30 instructions, making it unusable for performance-critical code. I think we have improved code since then, and compiler got better at abstracting away the penalty, so it might be worth to re-consider. Another consideration was that the storage order of `SymmetricTensor` was (probably still is) such that one does unnecessary accesses to the stored geometry arrays when only the diagonal is accessed, as is needed e.g. when you do not need the full Hessian but only the Laplacian. So I think those should be carefully benchmarked to see whether these arguments are irrelevant by now - in that case, we should definitely go for the simpler data structure that already exists.